gxa/src/main/java/uk/ac/ebi/atlas/experimentpage/differential/DifferentialPathwaysComparisonService.java

```diff
@@ -0,0 +1,133 @@
+        Map<Contrast, List<String>> comparisonsPathwaysMap = new HashMap<>();
+
+        String experimentAccession = experiment.getAccession();
+        List<Contrast> comparisons = configurationTrader.getMicroarrayExperimentConfiguration(experimentAccession).getContrasts();
+
+        //fetch pathwaysId from file for each comparison
+        for (Contrast comparison : comparisons) {
+            List<String> pathwaysValues = fetchPathwaysFromFile(experimentAccession, comparison);
+
+            comparisonsPathwaysMap.put(comparison, pathwaysValues);
+        }
+
+        return comparisonsPathwaysMap;
+    }
+
+    private List<String> fetchPathwaysFromFile (String experimentAccession, Contrast comparison) {
+        Map<String, Double> result = new HashMap<>();
+
+        List<String[]> lines = dataFileHub.getReactomePathwaysCFiles(experimentAccession, comparison.getId()).reactomePathways.get().readAll();
+
+        for(int i = 1; i < lines.size(); i++) {
+            String[] strings = lines.get(i);
+            String pathway = strings[0];
+            Double pvalue = Double.valueOf(strings[4]);
+
+            if (pvalue < 0.05) {
+                result.put(pathway, pvalue);
+            }
+        }
+
+        Map<String, Double> ordered = result.entrySet()
+                .stream()
+                .sorted(Map.Entry.comparingByValue())
+                .limit(10)
+                .collect(toMap(Map.Entry::getKey,
+                        Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+
+        return new ArrayList<>(ordered.keySet());
+    }
+
+    public Map<Contrast, SetMultimap<String, Pair<String, DifferentialExpression>>> constructPathwaysByComparison (E experiment,
+                                                                                                                   K preferences) {
+        //Read pathwaysId per comparison from file ({comparison}-reactome.gsea.tsv)
+        Map<Contrast, List<String>> contrastPathwaysListMap = getPathwaysListPerComparison(experiment);
+        SetMultimap<String, Pair<String, DifferentialExpression>> pathwayGeneExpressionMap = HashMultimap.create();
+
+        Map<Contrast, SetMultimap<String, Pair<String, DifferentialExpression>>> result = new HashMap<>();
+
+        //For each comparison and each pathway get differential profiles (geneIds and differential expressions)
+        for (Map.Entry<Contrast, List<String>> comparison : contrastPathwaysListMap.entrySet()) {
+            for (String pathwayId : comparison.getValue()) {
+                preferences.setGeneQuery(SemanticQuery.create(pathwayId));
+                R requestContext = differentialRequestContextFactory.create(experiment, preferences);
+
+                DifferentialProfilesList<P> differentialProfiles = profilesHeatMap.fetch(requestContext);
+
+                if (!differentialProfiles.isEmpty()) {
+                    for (int i = 0; i < differentialProfiles.getTotalResultCount(); i++) {
```

Comment:
I think this is a bug, and that your code will sometimes break here.

differentialProfiles.getTotalResultCount() can be more than the amount of elements in the list. We use it to display on the experiment page something like "Showing 50 of 9871 genes found".

Please use a foreach loop, and I would suggest it to be the default practice when you don't use the index for anything:
```
for(P profile: differentialProfiles){
//
}
```

Reply:
Ok, you are right. I didn't think that this can be higher to thousands, so I will change it.